Native compilation for bsb (PR 3/3)

bsansouci opened the pull request:
Hey @bobzhang sorry I took so long!

This simple first step should work well with only 1 target (it doesn't support multiple targets). I can build `BetterError` with it.

Here's the breakdown of what's going on here.

There are very few internal changes, and a bunch of additions. I made sure that the js flow wasn't hampered by anything that I did.
Here are the small internal changes:
- `package_specs_from_json` inside `bsb_config_parse` became `package_specs_and_buildable_targets_from_json` because I need that information returned. I update all the callsites.
- the variable `bsdep` that's exposed globally inside the build.ninja file is now called `bsb_helper` (that's what it was before, just different name) beause I'm using `bsb_helper` as a packer/linker since it seemed like a good place to put that code. I could also make that a standalone exec but I thought it fit well under the "helper" umbrella.
- `Bsb_ninja.handle_file_groups` now takes `root_project_compilation_kind` and `main_module_name` because I need those for added the packing/linking step.
- bsb_helper is now a packer/linker :)
- `bsb` now accept another internal cmdline arg called `-build-library` which helps when building external deps recursively. I call it like `bsb -build-library js,native,bytecode` to tell the external dep to build using those compilers. This allows the root project to tell its external dep how to build.
- `Ast_extract` has a bunch of small changes to give me the freedom to build the dependency graph and check stuff about it, and sort it without allocating too much. I didn't write new functions, just simply shifting arguments around and exposing helpers.
- I ouput another file, a `.mlast_simple` file, when `bsc` is ran, which is just a binary dump of the AST that `ocamlc`/`ocamlopt` understand

There is a big chunk of added commands inside `bsb_ninja` to support outputting different artifacts and to support the linking/packing step.

The reason I turned `bsb_helper` into a custom packer/linker was because that way we don't need to regenerate the build.ninja on any file change, just on file addition (which I'm guessing is what you do). It helps keep the dependency graph dynamic, as it's recomputed when `bsb_helper` is ran at packing/linking time.

Right now this just takes the first target and builds according to that. It also doesn't respect the dev/lib group nor does it respect the export field for libraries. The reason behind this is that I felt that it would overlap with module alias, and I know you're working on that. I have another branch which supports outputting build artifacts in special folders like `js_artifacts` and all, but it created a lot of duplicated artifacts (`mlast` mostly) and I didn't feel like the flow made sense, so I reduced the scope to this.

cc @chenglou @schmavery

chenglou commented:
Example bsconfig addition:

```json
  "buildable-targets": [
    {
      "kind": "bytecode",
      "entrypoint": "path/to/entrypoint",
    },
  ]
```

bobzhang commented on jscomp/bsb/bsb_gen.ml:
```diff
@@ -94,7 +97,7 @@ let output_ninja
           "src_root_dir", cwd (* TODO: need check its integrity -- allow relocate or not? *);
           "bsc", bsc ;
-          "bsdep", bsdep;
+          "bsb_helper", bsb_helper;
```
this is correct, it is called `bsdep` due to legacy and my laziness

bsansouci replied:
haha

bobzhang commented:
some high-level suggestions to help we land it sooner:

1. It would be nice to split into small patches : )
2. It would make sense to support `ocamlopt` first, we can leave `ocamlc` later, the reason is that `ocamlc` has different dependency model than `ocamlopt`, you can do signficantly better in that regard
3. It is possible to share the same format `.mlast`? I would like to document its header, currently its format is `....header-meta-data....+simple-ast`, currently you seem to generate it twice, it also makes dependency graph larger
4. Would you document it a bit  more how do you do the link? do you re-use info from `.d` or re-parse the source code

bsansouci commented:
I currently generate the build artifacts in the same place where you generate yours for simplicity. This one reason why this patch doesn't work with multiple targets (like `cmi`s will collide as we discussed).

Yes I do agree that generating those two files isn't optimal. I'm not entirely sure how to bypass this because `ocamlc`/`ocamlopt` only understand simple binary AST. Maybe if we patched refmt to support that format then we could invoke the compiler with `-pp "refmt --parse bsb_binary -print binary"`, but that might have performance implications (whereas here we mostly have memory concerns because we're duplicating data).

1) Sure! What about I push the parsing bsbconfig part first, then the changes to bsb_helper, then the rest?
2) You know much better than I do how the compiler works, how does `ocamlc` differ from `ocamlopt` in terms of dependency model? My approach was actually to prioritize `ocamlc` simply because it's faster than `ocamlopt`. Also I thought that I would start with a less-than-optimal PR and we'd work together to optimize for each compiler afterwards. This allows people to try those features early on and give us some feedback as to how they use it, or what bugs they encounter.
3) I'd love to but I'm not entirely sure how to (See my answer above)
4) Yes I use the mlast files to recompute the dependency graph because it wasn't clear to me that parsing `.d` files would be faster. Does bsb generate a binary dependency graph that I could just load up? That seems like it'd be the fastest.

bobzhang commented:
hmm, it is not that bad to keep the original simple format (a little loss of perf)
1. looks good 
2. ocamlc does not do any cross module compilation, while ocamlopt and bsc does, the latter are similar -- split it into two parts also help make the patch smaller
3. let's keep it as you have now
4. I suggest you should parsing `.d`, re-parse source files are not scalable, suppose you have 100,000 loc in a single file while `.d` maybe only two lines. maybe you can tweak `.d` generation to make your life easier. as long as generated `.d` is understandable by ninja, your parser does not need to be too complicated -- this may take some time than what you have now, but it is worthwhile even for MVP

bsansouci commented:
Hey @bobzhang I just updated this PR. This should be the 3rd one of the series :)

I'd love some help making this faster. Right now running on immutable-re (for example) we get 3.5s for js, between 4s and 5s for bytecode and around 7s for native.

One open question is which version of ocaml should we use? Should we assume the user has ocaml installed? Or should we simply use the ocaml that's vendored in bucklescript?

bobzhang commented:
The simplest solution would be using our vendored ocaml compiler, 
my current interest is not focused on native, let me know what you think on your side.
So when this PR land, we can build some simple examples in both native and JS port? If so, would you write some docs(marked as experimental) for users? Thanks! Let me know when it is ready for review

reply@reply.github.com At: 04/13/17 02:13:59" data-digest="From: reply@reply.github.com At: 04/13/17 02:13:59" style="">
From: reply@reply.github.com At: 04/13/17 02:13:59
To: bucklescript@noreply.github.com
Cc: Hongbo Zhang (BLOOMBERG/ 731 LEX), mention@noreply.github.com
Subject: Re: [bloomberg/bucklescript] Native compilation for bsb (#1338)


Hey @bobzhang I just updated this PR. This should be the 3rd one of the series :) 
I'd love some help making this faster. Right now running on immutable-re (for example) we get 3.5s for js, between 4s and 5s for bytecode and around 7s for native. 
One open question is which version of ocaml should we use? Should we assume the user has ocaml installed? Or should we simply use the ocaml that's vendored in bucklescript?  
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

bsansouci commented:
Ok I updated the PR to call into the shell `which ocamlc` and use that if it's defined. If not, we fallback to using our local one. For Windows I default to using the local one. I have NOT tested it on windows (though I can do that, I do have a machine). You did say that windows native support would be tough so I'm guessing this will not simply work.

I'd love to add examples. Do you want me to create a special `examples` folder? I can also just make another repo and point to it. Or I could use some existent project; I think @chenglou mentioned he wouldn't mind switching BetterErrors to this once we got something working, and that's one of the projects I've been using to test bib-native lol).

By the way, I looked over bsb_ninja and realized that the native flow is super similar to the JS one, but there are just different rules. I kept it inlined until now but I would be ok with moving all that stuff into another file, so I don't invade the js flow too much. There might be _some_ duplicated logic but I think it would be nicer for the future. Tell me what you think (you can look at the bsb_ninja file to get a sense of the updates).

I also add a function called `get_ocaml_dir` which depends on `bsb_unix`, adding a dependency to `bsb_build_util`. I know you don't like adding deps like those but I thought it wasn't too much of a stretch here. Alternatively I could simply inline that function where it's used (`Bsb_main` and `Bsb_helper_linker`/`Bsb_helper_packer`)

bsansouci commented:
New updates! I moved all the complex intrusive logic from `Bsb_ninja` to `Bsb_ninja_native` which contains currently a bunch of duplicated logic. I think this is better than bloating `Bsb_ninja` for two reasons:
- this is less intrusive for you to maintain the main JS flow. If I mashed everything to be inside Bsb_ninja the file gets pretty big and just complex, making the task of maintaining the main JS flow harder and we don't want that.
- this is more future proof. We will certainly end up with different ways of generating the ninja file as each target has its own optimizations that can be done. For example for native we can afford to make a general dependency tree (across external deps) and compile only what's used.

I also sneakily added a new field to the bsconfig which is `ocaml-flags`. I didn't bother documenting it because I think we'll probably want to give this a bit more thought (For example, there are compile flags and linker flags, and you might want to give different ones to different targets etc...). Right now this is used to pass the same warnings as the ones bsc uses. This fixes the issue that when building to JS you would get a lot more warnings than to native.

Tell me what you think!

bobzhang commented:
comments inlined
reply@reply.github.com At: 04/17/17 03:04:41" data-digest="From: reply@reply.github.com At: 04/17/17 03:04:41" style="">
From: reply@reply.github.com At: 04/17/17 03:04:41
To: bucklescript@noreply.github.com
Cc: Hongbo Zhang (BLOOMBERG/ 731 LEX), mention@noreply.github.com
Subject: Re: [bloomberg/bucklescript] Native compilation for bsb (PR 3/3) (#1338)


New updates! I moved all the complex intrusive logic from Bsb_ninja to Bsb_ninja_native which contains currently a bunch of duplicated logic. I think this is better than bloating Bsb_ninja for two reasons: 
*this is less intrusive for you to maintain the main JS flow. If I mashed everything to be inside Bsb_ninja the file gets pretty big and just complex, making the task of maintaining the main JS flow harder and we don't want that.
* 
*this is more future proof. We will certainly end up with different ways of generating the ninja file as each target has its own optimizations that can be done. For example for native we can afford to make a general dependency tree (across external deps) and compile only what's used. 
> This  looks good to me
I also sneakily added a new field to the bsconfig which is ocaml-flags. I didn't bother documenting it because I think we'll probably want to give this a bit more thought (For example, there are compile flags and linker flags, and you might want to give different ones to different targets etc...). Right now this is used to pass the same warnings as the ones bsc uses. This fixes the issue that when building to JS you would get a lot more warnings than to native.
> can you elaborate it a bit? btw, ocamlc and ocamlopt would accept different flags too 
Tell me what you think!  
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

bsansouci commented:
Sure, I added `let ocaml_flags = ["-w"; "-40+6+7+27+32..39+44+45"]` which I then used for bcc and for ocamlc/ocamlopt. Before doing this we were just calling without any of those, so compiling to JS would warn a lot more than compiling to native lol.

What other flags would you want to add?

bsansouci commented:
I updated to having a `warnings` field used by ocamlc/ocamlopt/bsc and removed the warnings from the `bsc_flags`. This should not break anyone. One thing that could happen is if someone overwrite the `bsc-flags` they might see the warnings appear again

bobzhang commented on jscomp/bsb/bsb_default.ml:
```diff
@@ -44,12 +44,12 @@
     "-no-alias-deps";
     "-color"; "always" ;
-    "-w"; "-40+6+7+27+32..39+44+45"
-
```
why?

bsansouci replied:
I added a new field to the bsconfig called `warnings` which takes that ^ values by default, and I added that flag to all commands which uses `bsc_flags`. All this to say that the compiler gets the same flags by default, so no behavioral changes. The idea is that you'll probably want the same warnings on native than on JS. (see my last two comments on this PR)

bsansouci commented:
Thanks a lot for reviewing this today.

1) Ok! Here's the a potential PR breakdown which will be easier to chew. This is just to give you an idea of what I'm thinking. Hope it help.
  a) make the change to bsb_helper to use `Bsb_build_util.get_ocaml_dir` (and implement it) + add `Bsb_unix.run_command_capture_stdout` (so modify `bsb_helper_linker`, `bsb_helper_packer`, `bsb_helper_main`)
  b) Update `Bsb_config_parse.package_specs_and_entries_from_bsconfig` to return the tuple I need `(package_specs, entries)` and add `warnings` (since it's in the same file) and `warn-errors`. That is, if you think it's a good idea.
  c) Modify `bsb_gen` to check the project entry point and call the JS ninja generation code (and fail when you try to compile to native/bytecode).
  d) Add the install calls for `.a`, `.cma`, `.cmxa` files inside `bsb_main`. (So bsb copies those files from `lib/bs` to `lib/ocaml`) + add the `external_deps_for_linking` which is a list of external libs (used for the linking/packing process) to `regenerate_ninja`.
  e) modify `js_implementation` to output a `mlast_simple` which is a the simple binary AST to be consumed by ocamlc/ocamlopt
  
  f) After all of these changes ^, no regression should be observed. Bsb should work flawlessly when building to JS.
  Then finally add the build rules in `bsb_rules` and add the new file `bsb_ninja_native` which generates the `ninja.build`.
  
2) Will do!

3) For tests, I'll add them.
For generating a jbuilder file, I'll look into it. I think it'd be outside of the scope of this PR but worth considering jbuilder.

bsansouci commented:
I'm gonna close this PR. I updated my fork so that master was where my changes were (to simplify things), so I won't be updating this branch anymore.
I'll use this PR to track progress as I go along though.
